# Cohort Analysis — Sadigov Edalet

## 1. Data Overview və Data Cleaning

### Data Overview


In [370]:
import numpy as np
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
from scipy import stats as stt
from plotly.subplots import make_subplots
import ipywidgets as widgets
from ipywidgets import HBox, VBox, HTML
from IPython.display import display

In [279]:
sales = pd.read_csv('sales.csv',low_memory=False)
sales

,order_id,order_date,status,item_id,sku,qty_ordered,price,discount_amount,category,payment_method,...,Customer Since,SSN,Phone No.,Place Name,County,City,State,Zip,Region,User Name
0,100354678,01-10-2020,received,574772,oasis_Oasis-064-36,21,89.9,0.0,Men's Fashion,cod,...,8/22/2006,627-31-5251,405-959-1129,Vinson,Harmon,Vinson,OK,73571,South,jwtitus
1,100354678,01-10-2020,received,574774,Fantastic_FT-48,11,19.0,0.0,Men's Fashion,cod,...,8/22/2006,627-31-5251,405-959-1129,Vinson,Harmon,Vinson,OK,73571,South,jwtitus
2,100354680,01-10-2020,complete,574777,mdeal_DMC-610-8,9,149.9,0.0,Men's Fashion,cod,...,8/22/2006,627-31-5251,405-959-1129,Vinson,Harmon,Vinson,OK,73571,South,jwtitus
3,100354680,01-10-2020,complete,574779,oasis_Oasis-061-36,9,79.9,0.0,Men's Fashion,cod,...,8/22/2006,627-31-5251,405-959-1129,Vinson,Harmon,Vinson,OK,73571,South,jwtitus
4,100367357,13-11-2020,received,595185,MEFNAR59C38B6CA08CD,2,99.9,0.0,Men's Fashion,cod,...,8/22/2006,627-31-5251,405-959-1129,Vinson,Harmon,Vinson,OK,73571,South,jwtitus
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
286387,100562365,30-09-2021,paid,905179,APPCHA5AF14939B8F8A,2,4419.9,0.0,Appliances,Easypay,...,3/21/2007,613-87-0361,212-772-7404,Rushville,Yates,Rushville,NY,14544,Northeast,bklatham
286388,100562376,30-09-2021,cod,905191,MEFCOT5A8D1E973B886,2,39.9,0.0,Men's Fashion,cod,...,10/24/2011,101-02-1040,229-817-9451,Lawrenceville,Gwinnett,Lawrenceville,GA,30044,South,bmbrunetti
286389,100562383,30-09-2021,cod,905200,WOFVAL59D5EA84167F9-M,2,40.0,0.0,Women's Fashion,cod,...,7/25/1987,399-31-7238,252-414-8396,Durham,Durham,Durham,NC,27701,South,fngiusti
286390,100562384,30-09-2021,cod,905202,WOFNIG5B4D7EB0E9FDD-L,2,49.9,0.0,Women's Fashion,cod,...,7/25/1987,399-31-7238,252-414-8396,Durham,Durham,Durham,NC,27701,South,fngiusti


In [280]:
sales.columns = (
    sales.columns
    .str.strip()
    .str.lower()
    .str.replace(' ', '_')
    .str.replace('.', '', regex=False)
)
print(sales.columns.tolist())

['order_id', 'order_date', 'status', 'item_id', 'sku', 'qty_ordered', 'price', 'discount_amount', 'category', 'payment_method', 'bi_st', 'cust_id', 'year', 'month', 'ref_num', 'name_prefix', 'first_name', 'middle_initial', 'last_name', 'gender', 'age', 'full_name', 'e_mail', 'customer_since', 'ssn', 'phone_no', 'place_name', 'county', 'city', 'state', 'zip', 'region', 'user_name']


In [281]:
print(f'Column count  :{sales.shape[1]}' )
print(f'Row count     :{sales.shape[0]}' )

Column count  :33
Row count     :286392


In [282]:
sales.info()

<class 'pandas.DataFrame'>
RangeIndex: 286392 entries, 0 to 286391
Data columns (total 33 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   order_id         286392 non-null  str    
 1   order_date       286392 non-null  str    
 2   status           286392 non-null  str    
 3   item_id          286392 non-null  int64  
 4   sku              286392 non-null  str    
 5   qty_ordered      286392 non-null  int64  
 6   price            286392 non-null  float64
 7   discount_amount  286392 non-null  float64
 8   category         286392 non-null  str    
 9   payment_method   286392 non-null  str    
 10  bi_st            286392 non-null  str    
 11  cust_id          286392 non-null  int64  
 12  year             286392 non-null  int64  
 13  month            286392 non-null  str    
 14  ref_num          286392 non-null  int64  
 15  name_prefix      286392 non-null  str    
 16  first_name       286392 non-null  str    
 17  mi

In [283]:
sales.isnull().sum()

order_id           0
order_date         0
status             0
item_id            0
sku                0
qty_ordered        0
price              0
discount_amount    0
category           0
payment_method     0
bi_st              0
cust_id            0
year               0
month              0
ref_num            0
name_prefix        0
first_name         0
middle_initial     0
last_name          0
gender             0
age                0
full_name          0
e_mail             0
customer_since     0
ssn                0
phone_no           0
place_name         0
county             0
city               0
state              0
zip                0
region             0
user_name          0
dtype: int64

In [284]:
print(f'Unique orders           : {sales['order_id'].nunique()}')
print(f'Unique customer         : {sales['cust_id'].nunique()}')
print(f'Unique categories       : {sales['category'].nunique()}')
print(f'Unique regions          : {sales['region'].nunique()}')
print(f'Unique payment methods  : {sales['payment_method'].nunique()}')
sales.dtypes.to_frame('dtype')

Unique orders           : 201716
Unique customer         : 64248
Unique categories       : 15
Unique regions          : 4
Unique payment methods  : 13


,dtype
order_id,str
order_date,str
status,str
item_id,int64
sku,str
qty_ordered,int64
price,float64
discount_amount,float64
category,str
payment_method,str


In [285]:
missing = sales.isnull().sum()
missing_pct = (missing / len(sales) * 100).round(2)

In [286]:
print(f'--------missing_pct--------')
display(missing_pct)
print(f'--------missing--------')
display(missing)

--------missing_pct--------


order_id           0.0
order_date         0.0
status             0.0
item_id            0.0
sku                0.0
qty_ordered        0.0
price              0.0
discount_amount    0.0
category           0.0
payment_method     0.0
bi_st              0.0
cust_id            0.0
year               0.0
month              0.0
ref_num            0.0
name_prefix        0.0
first_name         0.0
middle_initial     0.0
last_name          0.0
gender             0.0
age                0.0
full_name          0.0
e_mail             0.0
customer_since     0.0
ssn                0.0
phone_no           0.0
place_name         0.0
county             0.0
city               0.0
state              0.0
zip                0.0
region             0.0
user_name          0.0
dtype: float64

--------missing--------


order_id           0
order_date         0
status             0
item_id            0
sku                0
qty_ordered        0
price              0
discount_amount    0
category           0
payment_method     0
bi_st              0
cust_id            0
year               0
month              0
ref_num            0
name_prefix        0
first_name         0
middle_initial     0
last_name          0
gender             0
age                0
full_name          0
e_mail             0
customer_since     0
ssn                0
phone_no           0
place_name         0
county             0
city               0
state              0
zip                0
region             0
user_name          0
dtype: int64

### Data Cleaning

In [287]:
sales.columns.to_list()

['order_id',
 'order_date',
 'status',
 'item_id',
 'sku',
 'qty_ordered',
 'price',
 'discount_amount',
 'category',
 'payment_method',
 'bi_st',
 'cust_id',
 'year',
 'month',
 'ref_num',
 'name_prefix',
 'first_name',
 'middle_initial',
 'last_name',
 'gender',
 'age',
 'full_name',
 'e_mail',
 'customer_since',
 'ssn',
 'phone_no',
 'place_name',
 'county',
 'city',
 'state',
 'zip',
 'region',
 'user_name']

In [288]:
columns = ['order_id', 'order_date', 'status', 'item_id', 'sku', 'qty_ordered',
       'price', 'discount_amount', 'category', 'payment_method',
       'cust_id', 'year', 'month', 'gender', 'age','city', 'region']


cleaned = sales[columns].copy()

cleaned.head()

,order_id,order_date,status,item_id,sku,qty_ordered,price,discount_amount,category,payment_method,cust_id,year,month,gender,age,city,region
0,100354678,01-10-2020,received,574772,oasis_Oasis-064-36,21,89.9,0.0,Men's Fashion,cod,60124,2020,Oct-20,F,43,Vinson,South
1,100354678,01-10-2020,received,574774,Fantastic_FT-48,11,19.0,0.0,Men's Fashion,cod,60124,2020,Oct-20,F,43,Vinson,South
2,100354680,01-10-2020,complete,574777,mdeal_DMC-610-8,9,149.9,0.0,Men's Fashion,cod,60124,2020,Oct-20,F,43,Vinson,South
3,100354680,01-10-2020,complete,574779,oasis_Oasis-061-36,9,79.9,0.0,Men's Fashion,cod,60124,2020,Oct-20,F,43,Vinson,South
4,100367357,13-11-2020,received,595185,MEFNAR59C38B6CA08CD,2,99.9,0.0,Men's Fashion,cod,60124,2020,Nov-20,F,43,Vinson,South


In [289]:
print(f'Total duplicated rows                           : {cleaned.duplicated().sum()}')
print(f'Order_id and item_id together duplicated rows   : {cleaned.duplicated(subset=['order_id','item_id']).sum()}')
print(f'Avg line-items per order_id                     : {sales.groupby('order_id').size().mean():.2f}')

Total duplicated rows                           : 0
Order_id and item_id together duplicated rows   : 0
Avg line-items per order_id                     : 1.42


**This shows that there is no duplicated rows items orders in the dataset**

In [290]:
cleaned['order_date'] = pd.to_datetime(cleaned['order_date'], format='%d-%m-%Y')
cleaned['order_date'].describe()

count                        286392
mean     2021-03-12 11:18:32.008715
min             2020-10-01 00:00:00
25%             2020-12-24 00:00:00
50%             2021-03-16 00:00:00
75%             2021-05-18 00:00:00
max             2021-09-30 00:00:00
Name: order_date, dtype: object

In [291]:
print(f'Date range : {cleaned["order_date"].min().date()} - {cleaned["order_date"].max().date()}')

Date range : 2020-10-01 - 2021-09-30


Turning `order_date` column to datetime

In [292]:
cleaned['status'].value_counts()

status
canceled          112166
complete           88968
received           51775
order_refunded     25713
refund              3820
cod                 2851
paid                 756
closed               176
payment_review        57
pending               48
processing            33
holded                25
pending_paypal         4
Name: count, dtype: int64

I want to see `status` columns components counts here

**Status-based decision:**

We classify `status` into two groups:

- **Successful:** `complete`, `received`, `paid`, `closed`
- **Unsuccessful/incomplete:** `canceled`, `refund`, `order_refunded`, `pending`, `processing`, `payment_review`, `holded`, `pending_paypal`

**Note on `cod`:** the `status='cod'` value (2,851 rows) represents an order still in courier. That means we still did not get the money and I did not count it as completed.

Cohort and revenue analysis focuses only on actual completed sales, excluding canceled and refunded orders.

In [293]:
successful = ['complete', 'received', 'paid', 'closed']

orders_cleaned = cleaned[cleaned['status'].isin(successful)].copy()

print(f'Before cleaning the `status` column     : {cleaned.shape[0]}')
print(f'After cleanin the `status` column       :{orders_cleaned.shape[0]}' )
print(f'Reduced row count                       :{cleaned.shape[0] - orders_cleaned.shape[0]}' )

Before cleaning the `status` column     : 286392
After cleanin the `status` column       :141675
Reduced row count                       :144717


In [294]:
orders_cleaned.isnull().sum()

order_id           0
order_date         0
status             0
item_id            0
sku                0
qty_ordered        0
price              0
discount_amount    0
category           0
payment_method     0
cust_id            0
year               0
month              0
gender             0
age                0
city               0
region             0
dtype: int64




- No duplicates found (full rows or order_id + item_id).
- order_date was converted to datetime.
- 141,675 of 286,392 rows (~49.5%) had successful statuses and were retained.
- 144,717 unsuccessful/incomplete rows were removed.
- No missing values remain in the cleaned dataset.

### Revenue

In [295]:
orders_cleaned['revenue'] = ((orders_cleaned['qty_ordered']* orders_cleaned['price']) - orders_cleaned['discount_amount']).round(2)
orders_cleaned[['qty_ordered', 'price', 'discount_amount', 'revenue']].head()

,qty_ordered,price,discount_amount,revenue
0,21,89.9,0.0,1887.9
1,11,19.0,0.0,209.0
2,9,149.9,0.0,1349.1
3,9,79.9,0.0,719.1
4,2,99.9,0.0,199.8


In [296]:
print(f'Checking for a minus revenue: {(orders_cleaned['revenue']<0).sum()}')

Checking for a minus revenue: 0


In [297]:
orders_cleaned['revenue'].agg(mean = 'mean',median = 'median', std = 'std', max = 'max', min = 'min').round(2)

mean       1211.7
median      250.0
std        2328.4
max       46679.6
min           0.0
Name: revenue, dtype: float64

In [298]:
orders_cleaned['revenue'].describe().round(2)

count    141675.0
mean       1211.7
std        2328.4
min           0.0
25%          99.8
50%         250.0
75%        1384.6
max       46679.6
Name: revenue, dtype: float64



- `revenue` was successfully created with no negative values.
- Mean revenue = 1211.7, while median =250.
- Since the mean is much higher than the median, revenue is right-skewed, driven by a small number of high-value order lines.

## 2. Monthly cohort analysis

### First order date

In [299]:
first_order = orders_cleaned.groupby('cust_id')['order_date'].min().reset_index()
first_order.columns = ['cust_id','first_order_date']
first_order.head()

,cust_id,first_order_date
0,4,2020-12-06
1,15,2020-10-08
2,20,2021-03-20
3,21,2021-02-03
4,23,2021-02-23


In [300]:
orders_cleaned = orders_cleaned.merge(first_order,on= 'cust_id',how = 'left')
orders_cleaned

,order_id,order_date,status,item_id,sku,qty_ordered,price,discount_amount,category,payment_method,cust_id,year,month,gender,age,city,region,revenue,first_order_date
0,100354678,2020-10-01,received,574772,oasis_Oasis-064-36,21,89.9,0.0,Men's Fashion,cod,60124,2020,Oct-20,F,43,Vinson,South,1887.9,2020-10-01
1,100354678,2020-10-01,received,574774,Fantastic_FT-48,11,19.0,0.0,Men's Fashion,cod,60124,2020,Oct-20,F,43,Vinson,South,209.0,2020-10-01
2,100354680,2020-10-01,complete,574777,mdeal_DMC-610-8,9,149.9,0.0,Men's Fashion,cod,60124,2020,Oct-20,F,43,Vinson,South,1349.1,2020-10-01
3,100354680,2020-10-01,complete,574779,oasis_Oasis-061-36,9,79.9,0.0,Men's Fashion,cod,60124,2020,Oct-20,F,43,Vinson,South,719.1,2020-10-01
4,100367357,2020-11-13,received,595185,MEFNAR59C38B6CA08CD,2,99.9,0.0,Men's Fashion,cod,60124,2020,Nov-20,F,43,Vinson,South,199.8,2020-10-01
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
141670,100562293,2021-09-29,paid,905066,MATBIZ5A7032285724F,2,24.5,0.0,Mobiles & Tablets,bankalfalah,115304,2021,Sep-21,F,59,Hasbrouck Heights,Northeast,49.0,2021-09-29
141671,100562293,2021-09-29,paid,905067,MATHUA5B54C16B346E2,2,15.0,0.0,Mobiles & Tablets,bankalfalah,115304,2021,Sep-21,F,59,Hasbrouck Heights,Northeast,30.0,2021-09-29
141672,100562338,2021-09-30,paid,905145,WOFFAS5B22B977EA817-M,2,99.9,0.0,Women's Fashion,bankalfalah,115317,2021,Sep-21,F,71,Mc Farland,Midwest,199.8,2021-09-30
141673,100562352,2021-09-30,paid,905161,BAGKEM5A703890E559C,2,57.2,0.0,Beauty & Grooming,bankalfalah,23611,2021,Sep-21,F,68,Acworth,Northeast,114.4,2021-09-30


### Monthly period

In [301]:
orders_cleaned['cohort_month']=orders_cleaned['first_order_date'].dt.to_period('M')
orders_cleaned['order_month'] = orders_cleaned['order_date'].dt.to_period('M')

orders_cleaned

,order_id,order_date,status,item_id,sku,qty_ordered,price,discount_amount,category,payment_method,...,year,month,gender,age,city,region,revenue,first_order_date,cohort_month,order_month
0,100354678,2020-10-01,received,574772,oasis_Oasis-064-36,21,89.9,0.0,Men's Fashion,cod,...,2020,Oct-20,F,43,Vinson,South,1887.9,2020-10-01,2020-10,2020-10
1,100354678,2020-10-01,received,574774,Fantastic_FT-48,11,19.0,0.0,Men's Fashion,cod,...,2020,Oct-20,F,43,Vinson,South,209.0,2020-10-01,2020-10,2020-10
2,100354680,2020-10-01,complete,574777,mdeal_DMC-610-8,9,149.9,0.0,Men's Fashion,cod,...,2020,Oct-20,F,43,Vinson,South,1349.1,2020-10-01,2020-10,2020-10
3,100354680,2020-10-01,complete,574779,oasis_Oasis-061-36,9,79.9,0.0,Men's Fashion,cod,...,2020,Oct-20,F,43,Vinson,South,719.1,2020-10-01,2020-10,2020-10
4,100367357,2020-11-13,received,595185,MEFNAR59C38B6CA08CD,2,99.9,0.0,Men's Fashion,cod,...,2020,Nov-20,F,43,Vinson,South,199.8,2020-10-01,2020-10,2020-11
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
141670,100562293,2021-09-29,paid,905066,MATBIZ5A7032285724F,2,24.5,0.0,Mobiles & Tablets,bankalfalah,...,2021,Sep-21,F,59,Hasbrouck Heights,Northeast,49.0,2021-09-29,2021-09,2021-09
141671,100562293,2021-09-29,paid,905067,MATHUA5B54C16B346E2,2,15.0,0.0,Mobiles & Tablets,bankalfalah,...,2021,Sep-21,F,59,Hasbrouck Heights,Northeast,30.0,2021-09-29,2021-09,2021-09
141672,100562338,2021-09-30,paid,905145,WOFFAS5B22B977EA817-M,2,99.9,0.0,Women's Fashion,bankalfalah,...,2021,Sep-21,F,71,Mc Farland,Midwest,199.8,2021-09-30,2021-09,2021-09
141673,100562352,2021-09-30,paid,905161,BAGKEM5A703890E559C,2,57.2,0.0,Beauty & Grooming,bankalfalah,...,2021,Sep-21,F,68,Acworth,Northeast,114.4,2021-09-30,2021-09,2021-09


In [302]:
num_of_cohort = orders_cleaned['cohort_month'].nunique()

print(f'Count of monthly cohorts: {num_of_cohort}')
print(sorted(orders_cleaned['cohort_month'].unique()))

Count of monthly cohorts: 12
[Period('2020-10', 'M'), Period('2020-11', 'M'), Period('2020-12', 'M'), Period('2021-01', 'M'), Period('2021-02', 'M'), Period('2021-03', 'M'), Period('2021-04', 'M'), Period('2021-05', 'M'), Period('2021-06', 'M'), Period('2021-07', 'M'), Period('2021-08', 'M'), Period('2021-09', 'M')]


###  Cohort summary

In [303]:
cohort_summary =orders_cleaned.groupby('cohort_month').agg(
    unique_orders = ('order_id','nunique'),
    unique_customers = ('cust_id','nunique'),
    total_revenue = ('revenue','sum')
).reset_index()

cohort_summary = cohort_summary.sort_values('cohort_month').reset_index(drop=True)
cohort_summary

,cohort_month,unique_orders,unique_customers,total_revenue
0,2020-10,7821,1822,12018612.68
1,2020-11,8042,2522,12880225.37
2,2020-12,35055,14020,73357598.27
3,2021-01,4132,2369,3757260.39
4,2021-02,3269,1412,2502826.00
5,2021-03,8358,4175,20497506.66
6,2021-04,12083,6968,22240078.73
7,2021-05,3263,2305,3173567.52
8,2021-06,5740,2839,12744787.94
9,2021-07,2127,1422,4496352.17


###  Cohort summary visualization

In [304]:
cohort_summary.columns

Index(['cohort_month', 'unique_orders', 'unique_customers', 'total_revenue'], dtype='str')

In [305]:
fig_order = px.bar(
    cohort_summary,
    x=cohort_summary['cohort_month'].astype(str),
    y = 'unique_orders',
    title = 'Monthly unique orders for each cohort',
    labels= {'x': 'Month of cohort','unique_orders':'Count of unique orders'} 
)
fig_order.show()

In [306]:
fig_customer = px.bar(
    cohort_summary,
    x=cohort_summary['cohort_month'].astype(str),
    y= 'unique_customers',
    title = 'Count of monthly customers for each cohort ',
    labels={'x': 'month of cohort' , 'unique_customers':'Unique customers'}
)

fig_customer.show()

In [307]:
fig_revenue = px.bar(
    cohort_summary,
    x=cohort_summary['cohort_month'].astype(str),
    y= 'total_revenue',
    title = 'Total revenue of each cohort',
    labels={'x': 'month of cohort' , 'total_revenue':'Total revenue'}
)

fig_revenue.show()



All three charts show the same pattern: 2020-12 is the clear peak, followed by a sharp drop in 2021-01 and 2021-02. A second, moderate wave appears in 2021-04 and 2021-06, while metrics remain low and stable after 2021-07.

This suggests the business relied heavily on one-time campaign effects, while organic growth remained weak.

###  Customer activity pivot table

In [308]:
customer_activity = orders_cleaned.pivot_table(
    index = 'cohort_month',
    columns='order_month',
    values= 'cust_id',
    aggfunc='nunique'
)

customer_activity

order_month,2020-10,2020-11,2020-12,2021-01,2021-02,2021-03,2021-04,2021-05,2021-06,2021-07,2021-08,2021-09
cohort_month,,,,,,,,,,,,
2020-10,1822.0,306.0,381.0,169.0,123.0,177.0,151.0,88.0,143.0,85.0,54.0,80.0
2020-11,NaN,2522.0,562.0,211.0,135.0,204.0,178.0,106.0,188.0,82.0,47.0,82.0
2020-12,NaN,NaN,14020.0,1011.0,458.0,632.0,675.0,328.0,559.0,262.0,168.0,250.0
2021-01,NaN,NaN,NaN,2369.0,159.0,110.0,113.0,71.0,81.0,65.0,43.0,48.0
2021-02,NaN,NaN,NaN,NaN,1412.0,123.0,96.0,53.0,53.0,33.0,27.0,27.0
2021-03,NaN,NaN,NaN,NaN,NaN,4175.0,521.0,131.0,232.0,83.0,49.0,79.0
2021-04,NaN,NaN,NaN,NaN,NaN,NaN,6968.0,254.0,260.0,94.0,63.0,73.0
2021-05,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2305.0,159.0,89.0,49.0,40.0
2021-06,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2839.0,145.0,66.0,99.0


The diagonal shows each cohort’s initial size. Values drop sharply afterward, indicating low customer retention and many customers not returning after the first month.

### Monthly cohort metrisc

In [309]:
monthly_coh_met = orders_cleaned.groupby(['cohort_month','order_month']).agg(
    total_revenue = ('revenue','sum'),
    customer_count_unique = ('cust_id','nunique')
).reset_index()

monthly_coh_met

,cohort_month,order_month,total_revenue,customer_count_unique
0,2020-10,2020-10,2350821.91,1822
1,2020-10,2020-11,993603.23,306
2,2020-10,2020-12,3707018.58,381
3,2020-10,2021-01,409483.66,169
4,2020-10,2021-02,294192.07,123
...,...,...,...,...
73,2021-07,2021-08,874378.13,95
74,2021-07,2021-09,154495.06,52
75,2021-08,2021-08,2134201.71,1173
76,2021-08,2021-09,55175.21,61


The table shows revenue and customer count for each cohort-month combination in long format

### ARPU


In [310]:
monthly_coh_met['arpu'] = monthly_coh_met['total_revenue']/ monthly_coh_met['customer_count_unique']

arpu_pivot = monthly_coh_met.pivot_table(
    index= 'cohort_month',
    columns='order_month',
    values='arpu'
)

arpu_pivot.round(2)

order_month,2020-10,2020-11,2020-12,2021-01,2021-02,2021-03,2021-04,2021-05,2021-06,2021-07,2021-08,2021-09
cohort_month,,,,,,,,,,,,
2020-10,1290.24,3247.07,9729.71,2422.98,2391.81,8904.13,8148.85,2320.00,4306.99,2232.63,1720.70,4427.93
2020-11,NaN,1800.34,6390.71,2092.49,2329.32,6524.34,5800.67,2254.34,4155.67,2483.75,1629.52,4003.13
2020-12,NaN,NaN,4078.35,1333.52,2040.06,8168.87,6462.67,1757.58,4498.16,1376.23,1160.99,2899.96
2021-01,NaN,NaN,NaN,1096.81,1061.54,1689.95,2251.12,1890.50,1778.66,2663.77,1055.95,1104.51
2021-02,NaN,NaN,NaN,NaN,971.80,2969.73,2422.55,2364.37,3511.03,2714.33,2015.82,2867.18
2021-03,NaN,NaN,NaN,NaN,NaN,3563.63,6137.49,2605.78,6195.63,2125.83,1140.33,5198.29
2021-04,NaN,NaN,NaN,NaN,NaN,NaN,2829.94,1348.66,6032.99,1623.52,1502.92,4967.69
2021-05,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1044.76,3038.56,1774.08,1186.03,1656.54
2021-06,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3810.63,5634.72,9294.04,5009.82


ARPU often increases over time despite low retention. This means fewer customers return, but those who stay tend to spend more per customer.

> Low retention, but high value among retained customers.

###  Aylıq cohort lifetime

In [311]:
orders_cleaned['cohort_lifetime'] = orders_cleaned['order_month'].astype(int) - orders_cleaned['cohort_month'].astype(int)


orders_cleaned

,order_id,order_date,status,item_id,sku,qty_ordered,price,discount_amount,category,payment_method,...,month,gender,age,city,region,revenue,first_order_date,cohort_month,order_month,cohort_lifetime
0,100354678,2020-10-01,received,574772,oasis_Oasis-064-36,21,89.9,0.0,Men's Fashion,cod,...,Oct-20,F,43,Vinson,South,1887.9,2020-10-01,2020-10,2020-10,0
1,100354678,2020-10-01,received,574774,Fantastic_FT-48,11,19.0,0.0,Men's Fashion,cod,...,Oct-20,F,43,Vinson,South,209.0,2020-10-01,2020-10,2020-10,0
2,100354680,2020-10-01,complete,574777,mdeal_DMC-610-8,9,149.9,0.0,Men's Fashion,cod,...,Oct-20,F,43,Vinson,South,1349.1,2020-10-01,2020-10,2020-10,0
3,100354680,2020-10-01,complete,574779,oasis_Oasis-061-36,9,79.9,0.0,Men's Fashion,cod,...,Oct-20,F,43,Vinson,South,719.1,2020-10-01,2020-10,2020-10,0
4,100367357,2020-11-13,received,595185,MEFNAR59C38B6CA08CD,2,99.9,0.0,Men's Fashion,cod,...,Nov-20,F,43,Vinson,South,199.8,2020-10-01,2020-10,2020-11,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
141670,100562293,2021-09-29,paid,905066,MATBIZ5A7032285724F,2,24.5,0.0,Mobiles & Tablets,bankalfalah,...,Sep-21,F,59,Hasbrouck Heights,Northeast,49.0,2021-09-29,2021-09,2021-09,0
141671,100562293,2021-09-29,paid,905067,MATHUA5B54C16B346E2,2,15.0,0.0,Mobiles & Tablets,bankalfalah,...,Sep-21,F,59,Hasbrouck Heights,Northeast,30.0,2021-09-29,2021-09,2021-09,0
141672,100562338,2021-09-30,paid,905145,WOFFAS5B22B977EA817-M,2,99.9,0.0,Women's Fashion,bankalfalah,...,Sep-21,F,71,Mc Farland,Midwest,199.8,2021-09-30,2021-09,2021-09,0
141673,100562352,2021-09-30,paid,905161,BAGKEM5A703890E559C,2,57.2,0.0,Beauty & Grooming,bankalfalah,...,Sep-21,F,68,Acworth,Northeast,114.4,2021-09-30,2021-09,2021-09,0


In [312]:
print(f'Min lifetime                :{orders_cleaned['cohort_lifetime'].min()}' )
print(f'Maks lifetime               :{orders_cleaned['cohort_lifetime'].max()}' )
print(f'Count of negative lifetime  :{(orders_cleaned['cohort_lifetime'] < 0).sum()}' )

Min lifetime                :0
Maks lifetime               :11
Count of negative lifetime  :0


cohort_lifetime ranges from 0 to 11, with no negative values. This confirms that first_order_date was calculated correctly.

### ARPU heatmap

In [313]:
lifetime_metrics = orders_cleaned.groupby(['cohort_month','cohort_lifetime']).agg(
    total_revenue = ('revenue','sum'),
    unique_cust = ('cust_id', 'nunique')
).reset_index()

lifetime_metrics['arpu'] = lifetime_metrics['total_revenue']/ lifetime_metrics['unique_cust']


In [314]:

arpu_lifetime_pivot = lifetime_metrics.pivot_table(
    index= 'cohort_month',
    columns= 'cohort_lifetime',
    values= 'arpu'
)

arpu_lifetime_pivot.round(2)

cohort_lifetime,0,1,2,3,4,5,6,7,8,9,10,11
cohort_month,,,,,,,,,,,,
2020-10,1290.24,3247.07,9729.71,2422.98,2391.81,8904.13,8148.85,2320.00,4306.99,2232.63,1720.70,4427.93
2020-11,1800.34,6390.71,2092.49,2329.32,6524.34,5800.67,2254.34,4155.67,2483.75,1629.52,4003.13,NaN
2020-12,4078.35,1333.52,2040.06,8168.87,6462.67,1757.58,4498.16,1376.23,1160.99,2899.96,NaN,NaN
2021-01,1096.81,1061.54,1689.95,2251.12,1890.50,1778.66,2663.77,1055.95,1104.51,NaN,NaN,NaN
2021-02,971.80,2969.73,2422.55,2364.37,3511.03,2714.33,2015.82,2867.18,NaN,NaN,NaN,NaN
2021-03,3563.63,6137.49,2605.78,6195.63,2125.83,1140.33,5198.29,NaN,NaN,NaN,NaN,NaN
2021-04,2829.94,1348.66,6032.99,1623.52,1502.92,4967.69,NaN,NaN,NaN,NaN,NaN,NaN
2021-05,1044.76,3038.56,1774.08,1186.03,1656.54,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2021-06,3810.63,5634.72,9294.04,5009.82,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [315]:
arpu_lifetime_pivot_copy = arpu_lifetime_pivot.copy()
arpu_lifetime_pivot_copy.index = arpu_lifetime_pivot_copy.index.astype(str)

In [316]:
fig_heatmap = px.imshow(arpu_lifetime_pivot_copy,
                        text_auto='.0f',
                        color_continuous_scale= 'blues',
                        aspect='auto',
                        title='Monthly cohort ARPU lifetime heatmap',
                        labels=dict(x='Cohort lifetime (Month)', y='Month of cohort', color='ARPU'))

fig_heatmap.update_xaxes(side='top')
fig_heatmap.show()

## 3. Weekly retention analysis and final summary

### Weekly cohort

In [317]:
orders_cleaned['order_week'] = (
    orders_cleaned['order_date']
    - pd.to_timedelta(orders_cleaned['order_date'].dt.weekday, unit='D')
).dt.normalize()

orders_cleaned['first_purchase_week'] = (
    orders_cleaned['first_order_date']
    - pd.to_timedelta(orders_cleaned['first_order_date'].dt.weekday, unit='D')
).dt.normalize()

orders_cleaned[['order_date', 'order_week', 'first_order_date', 'first_purchase_week']].head()

,order_date,order_week,first_order_date,first_purchase_week
0,2020-10-01,2020-09-28,2020-10-01,2020-09-28
1,2020-10-01,2020-09-28,2020-10-01,2020-09-28
2,2020-10-01,2020-09-28,2020-10-01,2020-09-28
3,2020-10-01,2020-09-28,2020-10-01,2020-09-28
4,2020-11-13,2020-11-09,2020-10-01,2020-09-28


In [318]:
num_of_weekly_cohorts = orders_cleaned['first_purchase_week'].nunique()

print(f'Total count of weekly cohort: {num_of_weekly_cohorts}')

Total count of weekly cohort: 53


- `order_week`is first day of order week for each customer
- `first_purchase_week` is first day of the week which customer purchased in that week

### Weekly cohort lifetime

In [319]:
orders_cleaned['cohort_lifetime_weekly'] = (orders_cleaned['order_week'] - orders_cleaned['first_purchase_week']).dt.days // 7

orders_cleaned[['first_purchase_week', 'order_week', 'cohort_lifetime_weekly']].head()

,first_purchase_week,order_week,cohort_lifetime_weekly
0,2020-09-28,2020-09-28,0
1,2020-09-28,2020-09-28,0
2,2020-09-28,2020-09-28,0
3,2020-09-28,2020-09-28,0
4,2020-09-28,2020-11-09,6


In [320]:
print(f'Minimum : {orders_cleaned['cohort_lifetime_weekly'].min()}')
print(f'Maximum :{orders_cleaned['cohort_lifetime_weekly'].max()}')
print(f'Negative count: {(orders_cleaned['cohort_lifetime_weekly']<0).sum()}')


Minimum : 0
Maximum :52
Negative count: 0


###  Weekly active customers

In [321]:
weekly_active=orders_cleaned.groupby(
    ['first_purchase_week','cohort_lifetime_weekly'])['cust_id'].nunique().reset_index(name='active_customers')

weekly_active.head()

,first_purchase_week,cohort_lifetime_weekly,active_customers
0,2020-09-28,0,402
1,2020-09-28,1,23
2,2020-09-28,2,25
3,2020-09-28,3,25
4,2020-09-28,4,14


For each weekly cohort, the number of unique active customers was calculated for each lifetime week.

### Initial cohort size

In [322]:
cohort_size = weekly_active[weekly_active['cohort_lifetime_weekly']==0][['first_purchase_week','active_customers']]
cohort_size.columns = ['first_purchase_week', 'cohort_size']
cohort_size.head()

,first_purchase_week,cohort_size
0,2020-09-28,402
52,2020-10-05,441
103,2020-10-12,342
153,2020-10-19,397
202,2020-10-26,246


In [323]:
weekly_active

,first_purchase_week,cohort_lifetime_weekly,active_customers
0,2020-09-28,0,402
1,2020-09-28,1,23
2,2020-09-28,2,25
3,2020-09-28,3,25
4,2020-09-28,4,14
...,...,...,...
1364,2021-09-13,1,9
1365,2021-09-13,2,2
1366,2021-09-20,0,160
1367,2021-09-20,1,2


In [324]:
weekly_cohort_together = weekly_active.merge(cohort_size,on= 'first_purchase_week',how = 'left')
weekly_cohort_together

,first_purchase_week,cohort_lifetime_weekly,active_customers,cohort_size
0,2020-09-28,0,402,402
1,2020-09-28,1,23,402
2,2020-09-28,2,25,402
3,2020-09-28,3,25,402
4,2020-09-28,4,14,402
...,...,...,...,...
1364,2021-09-13,1,9,495
1365,2021-09-13,2,2,495
1366,2021-09-20,0,160,160
1367,2021-09-20,1,2,160


Each row was assigned its cohort’s size `cohort_size`

### Retention rate

In [325]:
weekly_cohort_together['retention_rate'] = (
    weekly_cohort_together['active_customers']/ weekly_cohort_together['cohort_size']
).round(5)

In [326]:
weekly_cohort_together.sample(5)

,first_purchase_week,cohort_lifetime_weekly,active_customers,cohort_size,retention_rate
484,2020-12-07,11,4,664,0.00602
854,2021-02-15,11,2,312,0.00641
1002,2021-03-22,14,13,1562,0.00832
753,2021-01-25,9,4,327,0.01223
29,2020-09-28,29,7,402,0.01741


The `retention_rate` column was created. At lifetime 0, retention is always 1.0 (100%), as expected, since it represents the cohort’s total size.

### Retention matrix and heatmap

In [327]:
retention_pivot = weekly_cohort_together.pivot_table(
    index='first_purchase_week',
    columns='cohort_lifetime_weekly',
    values='retention_rate'
)

retention_pivot

cohort_lifetime_weekly,0,1,2,3,4,5,6,7,8,9,...,42,43,44,45,46,47,48,49,50,52
first_purchase_week,,,,,,,,,,,,,,,,,,,,,
2020-09-28,1.0,0.05721,0.06219,0.06219,0.03483,0.05224,0.07214,0.06716,0.04975,0.03980,...,0.02736,0.02239,0.01493,0.00746,0.00995,0.00746,0.00746,0.01741,0.03483,0.00249
2020-10-05,1.0,0.07937,0.06122,0.03855,0.04989,0.04308,0.04989,0.05215,0.03401,0.04082,...,0.00680,0.01361,0.00680,0.00454,0.00680,0.00454,0.00907,0.01587,0.00907,NaN
2020-10-12,1.0,0.12865,0.06140,0.08772,0.08187,0.05263,0.05263,0.05556,0.09064,0.13158,...,0.02339,0.01754,0.00292,0.01170,0.00292,0.01754,0.05848,NaN,0.00292,NaN
2020-10-19,1.0,0.05542,0.07809,0.06045,0.03778,0.04030,0.04786,0.04030,0.08564,0.10076,...,0.00756,0.00252,0.01259,0.01259,0.01511,0.02015,NaN,0.00252,NaN,NaN
2020-10-26,1.0,0.11789,0.08537,0.06504,0.08943,0.05285,0.07317,0.11382,0.13415,0.06504,...,0.00813,0.00407,0.00813,0.00813,0.02846,0.00407,NaN,NaN,NaN,NaN
2020-11-02,1.0,0.11007,0.07260,0.06323,0.05855,0.04215,0.15222,0.17330,0.08197,0.05152,...,0.00468,0.00234,0.01171,0.03981,NaN,NaN,NaN,NaN,NaN,NaN
2020-11-09,1.0,0.05065,0.03873,0.04071,0.04568,0.10427,0.14697,0.05065,0.03078,0.02880,...,0.00298,0.00695,0.02979,0.00099,NaN,NaN,NaN,NaN,NaN,NaN
2020-11-16,1.0,0.04478,0.03648,0.03648,0.06799,0.08458,0.04809,0.01824,0.01990,0.01824,...,NaN,0.01327,0.00498,0.00166,NaN,NaN,NaN,NaN,NaN,NaN
2020-11-23,1.0,0.07500,0.03864,0.08409,0.12045,0.05227,0.04545,0.03864,0.02500,0.00455,...,0.01364,0.00909,0.00227,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [328]:
retention_pivot_copy = retention_pivot.copy()
retention_pivot_copy.index = retention_pivot_copy.index.astype(str)

fig_retention_heatmap = px.imshow(
    retention_pivot_copy * 100,
    color_continuous_scale='Reds',
    aspect='auto',
    title='Weekly retention heatmap (%)',
    labels=dict(x='Cohort lifetime (week)', y='First purchase week', color='Retention %')
)
fig_retention_heatmap.update_xaxes(side='top')
fig_retention_heatmap.show()

In [329]:
average_retention_by_week = (retention_pivot.mean() * 100).round(1)
average_retention_by_week.head(6)

cohort_lifetime_weekly
0    100.0
1      6.4
2      4.4
3      3.4
4      2.8
5      2.3
dtype: float64

The heatmap shows a sharp drop after lifetime 0. Average retention falls from 100% → 6.5% → 6.0% in weeks 0–2.

This means ~93–94% of customers do not return the following week, indicating very low retention and mostly one-time purchases.

### Churn rate

In [330]:
weekly_cohort_together.columns

Index(['first_purchase_week', 'cohort_lifetime_weekly', 'active_customers',
       'cohort_size', 'retention_rate'],
      dtype='str')

In [331]:
weekly_cohort_together['churn_rate'] = 1 - weekly_cohort_together['retention_rate']

churn_pivot = weekly_cohort_together.pivot_table(
    index='first_purchase_week',
    columns='cohort_lifetime_weekly',
    values='churn_rate'
)
weekly_cohort_together

,first_purchase_week,cohort_lifetime_weekly,active_customers,cohort_size,retention_rate,churn_rate
0,2020-09-28,0,402,402,1.00000,0.00000
1,2020-09-28,1,23,402,0.05721,0.94279
2,2020-09-28,2,25,402,0.06219,0.93781
3,2020-09-28,3,25,402,0.06219,0.93781
4,2020-09-28,4,14,402,0.03483,0.96517
...,...,...,...,...,...,...
1364,2021-09-13,1,9,495,0.01818,0.98182
1365,2021-09-13,2,2,495,0.00404,0.99596
1366,2021-09-20,0,160,160,1.00000,0.00000
1367,2021-09-20,1,2,160,0.01250,0.98750


In [ ]:
churn_pivot_copy = churn_pivot.copy()
churn_pivot_copy.index = churn_pivot_copy.index.astype(str)

fig_churn_heatmap = px.imshow(
    churn_pivot_copy * 100,
    color_continuous_scale='blues',
    aspect='auto',
    title='Period Churn Rate Heatmap (%)',
    labels=dict(x='Cohort lifetime (week)', y='First purchase week', color='Churn %')
)
fig_churn_heatmap.update_xaxes(side='top')
fig_churn_heatmap.show()

###  Final Summary

**Key observations:**

1. **Data volume and cleaning.** Of 286,392 order lines, only 141,675 (~49.5%) carry a successful status (`complete`, `received`, `paid`, `closed`); the other half were canceled, refunded, or still in progress (`cod`, `pending`, etc.). This points to a serious operational issue (payment, delivery) worth separate investigation.

2. **Cohort sizes are highly uneven.** The Dec-2020 cohort (14,020 customers) is 3–9x larger than every other month, pointing to a seasonal campaign. Size dropped sharply afterward (Jan/Feb 2021), with smaller secondary waves in Apr and Jun 2021.

3. **Retention is extremely low.** Average retention across weekly cohorts: week 0 → 100%, week 1 → ~6.5%, week 2 → ~6.0%, week 3 → ~3.4%. About 94% of customers never come back after their first purchase — a one-time-buyer pattern dominates.

4. **ARPU behaves counterintuitively.** The small group that does return spends **more** on average than first-time buyers (ARPU rises with lifetime in most cohorts). The loyal segment is small but high-value — focusing on it could drive disproportionate revenue.

5. **Churn is essentially irreversible.** Cumulative churn stabilizes above 90% after week 1 — once lost, a customer almost never returns.

**Business recommendations:**

- **First-week horror.** The biggest retention drop happens between week 0 and week 1. Targeted email/SMS and a second-purchase discount coupon should be tested in this time.
- **Campaign-driven customers.** Large one-off campaign cohorts (like Dec-2020) tend to retain worse — a dedicated win-back campaign should target this segment.
- **Loyal segment.** Rising ARPU over lifetime shows the small returning group is high-value — retain it with a VIP program or loyalty points.
- **Operational quality.** Half of all orders being canceled/refunded is a standalone problem (payment failure, stock, delivery delay) that needs its own operational analysis.


# New analysis

## Category analysis

In [333]:
category_performance = orders_cleaned.groupby('category').agg(
    revenue=('revenue', 'sum'),
    line_items=('category', 'size'),
    customers=('cust_id', 'nunique'),
    orders=('order_id', 'nunique')
).reset_index()

category_performance['aov'] = (category_performance['revenue'] / category_performance['orders']).round(1)

`Aov` = Average order value

In [334]:
orders_per_category = (
    orders_cleaned
    .groupby(['category', 'cust_id'])['order_id']
    .nunique()
    .reset_index(name='order_count')
)
orders_per_category['is_repeated'] = orders_per_category['order_count'] > 1

`İs_repeated`-> that is where there are more than 1 order count

In [335]:
repeat_rate = orders_per_category.groupby('category')['is_repeated'].mean() * 100
repeat_rate = repeat_rate.rename('repeat_customer_rate_%').reset_index()

category_performance = category_performance.merge(repeat_rate, on='category', how='left')
category_performance = category_performance.sort_values('revenue', ascending=False)

category_performance.round(1)

,category,revenue,line_items,customers,orders,aov,repeat_customer_rate_%
9,Mobiles & Tablets,81348729.7,26386,10354,25129,3237.2,40.3
0,Appliances,30940157.0,16580,7821,15015,2060.6,32.7
4,Entertainment,28930005.6,8492,3549,8344,3467.2,38.7
10,Others,6464628.9,7671,5132,7436,869.4,24.8
3,Computing,6223111.6,3584,2227,3244,1918.3,16.2
14,Women's Fashion,5759954.3,15502,6010,9171,628.1,22.7
8,Men's Fashion,3920387.9,21517,10595,14095,278.1,18.7
1,Beauty & Grooming,2242751.4,10484,4560,6497,345.2,20.4
13,Superstore,2171503.9,9135,2230,5004,434.0,33.6
6,Home & Living,1513333.8,7947,3619,4626,327.1,16.8


In [336]:
fig_category = px.bar(
    category_performance.sort_values('revenue'),
    x='revenue', y='category', orientation='h',
    color='repeat_customer_rate_%', color_continuous_scale='Blues',
    title='Category Revenue, coloured by Repeat-Customer Rate',
    labels={'revenue': 'Net revenue', 'category': '', 'repeat_customer_rate_%': 'Repeat rate (%)'}
)
fig_category.update_layout(height=520)
fig_category.show()

top_acquire = category_performance.sort_values('customers', ascending=False).iloc[0]
top_retain = category_performance.sort_values('repeat_customer_rate_%', ascending=False).iloc[0]

print(f'Acquisition lideri (ən çox unikal müştəri) : {top_acquire['category']} — {int(top_acquire['customers']):,} müştəri')
print(f'Retention lideri (ən yüksək repeat-rate)    : {top_retain['category']} — {top_retain['repeat_customer_rate_%']:.1f}%')

Acquisition lideri (ən çox unikal müştəri) : Men's Fashion — 10,595 müştəri
Retention lideri (ən yüksək repeat-rate)    : Mobiles & Tablets — 40.3%


Just 3 categories — `Mobiles & Tablets`, `Appliances`, and `Entertainment` — generate 82.3% of total revenue. `Mobiles & Tablets` alone is for 47.4%.

The remaining 12 categories contribute only ~18%, indicating high dependence on a small number of categories.

- `Men's Fashion` leads customer acquisition with 10,595 unique customers, despite its low AOV (278.1), making it a strong entry-point category.

- `Mobiles & Tablets` has the highest repeat-purchase rate (40.3%) and leads both revenue and retention.

- `School & Education` has the lowest repeat rate (8.9%), indicating mostly one-time purchases.

## RFM Segmentation

In [337]:
snapshot_date = orders_cleaned['order_date'].max() + pd.Timedelta(days=1)

rfm = orders_cleaned.groupby('cust_id').agg(
    last_purchase=('order_date', 'max'),
    frequency=('order_id', 'nunique'),
    monetary=('revenue', 'sum')
).reset_index()

rfm['recency'] = (snapshot_date - rfm['last_purchase']).dt.days

print(f'Snapshot Date: {snapshot_date.date()}')
rfm.head()

Snapshot Date: 2021-10-01


,cust_id,last_purchase,frequency,monetary,recency
0,4,2021-09-14,15,54747.45,17
1,15,2021-02-11,3,415.10,232
2,20,2021-07-29,6,47594.20,64
3,21,2021-02-03,1,126.00,240
4,23,2021-08-07,2,786.48,55


- `Recency` = Days since the last purchase
- `Frequency` = Number of unique orders
- `Monetary` = Total amount spent
- `Snapshot` Date = The day after the dataset’s last order date

In [338]:
rfm['R_score'] = pd.qcut(rfm['recency'].rank(method='first'), 5, labels=[5, 4, 3, 2, 1]).astype(int)
rfm['F_score'] = pd.qcut(rfm['frequency'].rank(method='first'), 5, labels=[1, 2, 3, 4, 5]).astype(int)
rfm['M_score'] = pd.qcut(rfm['monetary'].rank(method='first'), 5, labels=[1, 2, 3, 4, 5]).astype(int)

rfm['RFM_score'] = rfm['R_score'].astype(str) + rfm['F_score'].astype(str) + rfm['M_score'].astype(str)
rfm.head()

,cust_id,last_purchase,frequency,monetary,recency,R_score,F_score,M_score,RFM_score
0,4,2021-09-14,15,54747.45,17,5,5,5,555
1,15,2021-02-11,3,415.10,232,3,4,3,343
2,20,2021-07-29,6,47594.20,64,5,5,5,555
3,21,2021-02-03,1,126.00,240,3,1,1,311
4,23,2021-08-07,2,786.48,55,5,4,3,543


Scores are based on the dataset’s quintiles

- `Recency:` lower days = better → score 5
- `Frequency / Monetary:` higher = better → score 5
- `.rank(method='first')` helps qcut handle duplicate values without errors.

In [339]:
def assign_segment(row):
    r, f, m = row['R_score'], row['F_score'], row['M_score']
    if r >= 4 and f >= 4 and m >= 4:
        return 'Champions'
    elif r <= 2 and f >= 4 and m >= 4:
        return "Can't Lose Them"
    elif r >= 3 and f >= 3 and m >= 3:
        return 'Loyal Customers'
    elif r >= 4 and f <= 2:
        return 'New Customers'
    elif r <= 2 and f >= 3:
        return 'At Risk'
    elif r <= 2 and f <= 2 and m <= 2:
        return 'Hibernating/Lost'
    elif r >= 3 and f <= 3 and m <= 3:
        return 'Promising'
    else:
        return 'Potential Loyalists'

rfm['segment'] = rfm.apply(assign_segment, axis=1)
rfm['segment'].value_counts()

segment
Loyal Customers        8445
Potential Loyalists    7869
Promising              7586
Hibernating/Lost       6732
Champions              5214
Can't Lose Them        3625
At Risk                1500
New Customers          1203
Name: count, dtype: int64

Customers are segmented based on their RFM scores:

- `Champions`: High R, F, and M
- `Can’t Lose Them`: Low R, but high F and M
- `Loyal Customers`: High overall RFM scores
- `New Customers`: Recent but low-frequency buyers
- `At Risk`: Low R, but relatively high F
- `Hibernating/Lost`: Low R, F, and M
- `Promising`: Recent customers with moderate F and M
- `Potential Loyalists`: Remaining customers with loyalty potential.

In [340]:
segment_summary = rfm.groupby('segment').agg(
    customers=('cust_id', 'count'),
    revenue=('monetary', 'sum'),
    avg_frequency=('frequency', 'mean'),
    avg_recency=('recency', 'mean'),
).reset_index()

segment_summary['revenue_share_%'] = (segment_summary['revenue'] / segment_summary['revenue'].sum() * 100).round(1)
segment_summary['customer_share_%'] = (segment_summary['customers'] / segment_summary['customers'].sum() * 100).round(1)
segment_summary = segment_summary.sort_values('revenue', ascending=False).reset_index(drop=True)

segment_summary.round(1)

,segment,customers,revenue,avg_frequency,avg_recency,revenue_share_%,customer_share_%
0,Champions,5214,76333237.7,6.3,93.9,44.5,12.4
1,Can't Lose Them,3625,39061168.9,3.8,281.4,22.8,8.6
2,Loyal Customers,8445,35234322.1,2.2,150.6,20.5,20.0
3,Potential Loyalists,7869,15896677.7,1.1,225.5,9.3,18.7
4,Promising,7586,1893451.8,1.0,169.4,1.1,18.0
5,Hibernating/Lost,6732,1311651.8,1.0,290.4,0.8,16.0
6,New Customers,1203,1103507.8,1.0,101.2,0.6,2.9
7,At Risk,1500,833542.4,2.3,284.9,0.5,3.6


In [341]:
fig_rfm = px.scatter(
    segment_summary,
    x='avg_recency', y='avg_frequency',
    size='customers', color='revenue_share_%',
    text='segment',
    size_max=80,
    color_continuous_scale='RdYlGn',
    title='RFM Segment Map — Recency vs Frequency (bubble size = customers, colour = revenue share)',
    labels={'avg_recency': 'Avg Recency (days since last order) — lower is better',
            'avg_frequency': 'Avg Frequency (orders per customer)',
            'revenue_share_%': 'Revenue share (%)'}
)
fig_rfm.update_traces(textposition='top center', textfont=dict(size=11, color='black'))
fig_rfm.update_xaxes(autorange='reversed')  # low recency (recent buyers) shown on the right
fig_rfm.update_layout(height=560, coloraxis_colorbar_title='Revenue share (%)')
fig_rfm.show()

top2 = segment_summary.sort_values('revenue', ascending=False).iloc[:2]
print(f"Champions + Can't Lose Them: {top2['customer_share_%'].sum():.1f}% of customers "
      f"generate {top2['revenue_share_%'].sum():.1f}% of revenue.")

Champions + Can't Lose Them: 21.0% of customers generate 67.3% of revenue.


`Revenue is highly concentrated`: Only `Champions` (12.4% of customers) generate 44.5% of total revenue. `Champions + Can't Lose Them` represent just 21.0% of customers, but generate `67.3%` of revenue — nearly two-thirds of revenue depends on a small customer group.

The `largest` group is not always the most valuable: `Loyal Customers` make up 20.0% of customers and contribute 20.5% of revenue. Meanwhile, `Promising` (18.0%) and `Hibernating/Lost` (16.0%) together represent over one-third of customers but generate only 1.9% of revenue.

## Geographic Analysis

In [342]:
orders_cleaned.columns

Index(['order_id', 'order_date', 'status', 'item_id', 'sku', 'qty_ordered',
       'price', 'discount_amount', 'category', 'payment_method', 'cust_id',
       'year', 'month', 'gender', 'age', 'city', 'region', 'revenue',
       'first_order_date', 'cohort_month', 'order_month', 'cohort_lifetime',
       'order_week', 'first_purchase_week', 'cohort_lifetime_weekly'],
      dtype='str')

In [343]:
geography = orders_cleaned.groupby('region').agg(
    revenue = ('revenue','sum'),
    orders = ('order_id','nunique'),
    customer = ('cust_id','nunique')
).reset_index()

In [344]:
geography['aov'] = (geography['revenue']/ geography['orders']).round(2)
geography

,region,revenue,orders,customer,aov
0,Midwest,44955368.86,25986,11456,1729.98
1,Northeast,30152991.35,16230,7475,1857.86
2,South,64608907.33,33715,15615,1916.33
3,West,31950292.54,16944,7628,1885.64


In [345]:
orders_per_region = orders_cleaned.groupby(['region','cust_id'])['order_id'].nunique().reset_index(name = 'order_count')
orders_per_region

,region,cust_id,order_count
0,Midwest,4,15
1,Midwest,15,3
2,Midwest,54,1
3,Midwest,58,9
4,Midwest,64,77
...,...,...,...
42169,West,115202,1
42170,West,115218,1
42171,West,115251,1
42172,West,115268,1


In [346]:
orders_per_region['repeated'] = orders_per_region['order_count']>1

region_repeated = orders_per_region.groupby('region')['repeated'].mean()*100


In [347]:

region_repeated = region_repeated.rename('repeat_rate_%').reset_index().round(2)
region_repeated

,region,repeat_rate_%
0,Midwest,35.68
1,Northeast,34.82
2,South,35.27
3,West,35.49


In [348]:
geography = geography.merge(region_repeated,on='region')

In [349]:
geography['revenue_share_%'] = (geography['revenue'] / geography['revenue'].sum() * 100).round(2)
geography = geography.sort_values('revenue',ascending= False).reset_index(drop = True)

geography

,region,revenue,orders,customer,aov,repeat_rate_%,revenue_share_%
0,South,64608907.33,33715,15615,1916.33,35.27,37.64
1,Midwest,44955368.86,25986,11456,1729.98,35.68,26.19
2,West,31950292.54,16944,7628,1885.64,35.49,18.61
3,Northeast,30152991.35,16230,7475,1857.86,34.82,17.56


In [350]:
fig_geo = go.Figure()

fig_geo.add_trace(go.Bar(
    x=geography['region'], y=geography['revenue'],
    name='Revenue', marker_color='#2E86AB', yaxis='y1'
))
fig_geo.add_trace(go.Scatter(
    x=geography['region'], y=geography['repeat_rate_%'],
    name='Repeat Rate (%)', mode='lines+markers',
    marker=dict(color='#F26419', size=10), line=dict(width=3), yaxis='y2'
))

fig_geo.update_layout(
    title='Revenue by region, with Repeat- Rate Overlay',
    yaxis=dict(title='Net Revenue'),
    yaxis2=dict(title='Repeat rate (%)', overlaying='y', side='right', range=[0, 50]),
    height=480
)
fig_geo.show()

top_region = geography.iloc[0]
print(f"Top region by revenue: {top_region['region']} — {top_region['revenue_share_%']:.1f}% of total revenue")
print(f"Repeat-purchase rate range across regions: "
      f"{geography['repeat_rate_%'].min():.1f}% – {geography['repeat_rate_%'].max():.1f}%")

Top region by revenue: South — 37.6% of total revenue
Repeat-purchase rate range across regions: 34.8% – 35.7%


***Geographic Analysis*** 

Revenue is proportional to customer base, with no clear "Biggest":
- `South` leads with `37.6%` of revenue and `37.0%` of customers. Other regions show a similar pattern, meaning South’s lead comes from its larger customer base, not higher per-customer value.

Customer behavior is consistent across regions:
- `AOV` is similar `(1,730–1,916`), while `repeat-purchase` rates are nearly identical `(34.8%–35.7%)`. No meaningful regional behavioral difference is evident.

> Conclusion: Geographic targeting should focus mainly on market size, as no strong region-specific customer behavior was identified.

# Statistical Analysis

## stats

### Cust_statistic created for to make statistics easy

In [351]:
cust_statistic = orders_cleaned.groupby('cust_id').agg(
    total_revenue = ('revenue','sum'),
    num_orders = ('order_id','nunique'),
    total_quantity = ('qty_ordered','sum'),
    num_of_skus = ('sku','nunique'),
    region = ('region','first'),
    gender = ('gender', 'first'),
    payment_preference = ('payment_method',lambda x : x.mode().iat[0])
).reset_index()

- `groupby('cust_id').agg()` — Aggregates data to the customer level, since statistical tests compare customers, not line-items.
- `total_revenue=('revenue','sum')` — Total revenue generated by each customer.
- `n_orders=('order_id','nunique')` — Counts unique orders, avoiding duplicate line-items within the same order.
- `region / gender` — Uses the first value because these attributes are constant per customer.
- `payment_pre`f — Uses .mode() to find the customer’s most common payment method; .iat[0] selects the first if multiple modes exist.

In [352]:
cust_statistic['aov'] = (cust_statistic['total_revenue'] / cust_statistic['num_orders']).round(2)
cust_statistic['repeated'] = cust_statistic['num_orders'] > 1
cust_statistic.sample(5)

,cust_id,total_revenue,num_orders,total_quantity,num_of_skus,region,gender,payment_preference,aov,repeated
460,3816,6006.88,4,86,12,Northeast,F,cod,1501.72,True
26496,89230,9298.80,2,35,4,South,M,cod,4649.40,True
4990,46776,46629.70,4,8,3,Northeast,M,Easypay,11657.42,True
6995,60357,518.00,1,20,2,Northeast,M,cod,518.00,False
8893,63579,1006.20,1,3,1,West,M,Easypay,1006.20,False


in here  i wanted to see cust-level analysis. for my statistics 
- `Aov`- each customer’s average order value.
- `repeated` - Identifies whether a customer made more than one order

### Payment Method vs Repeat Purchase

In [353]:
cust_statistic.columns.to_list()

['cust_id',
 'total_revenue',
 'num_orders',
 'total_quantity',
 'num_of_skus',
 'region',
 'gender',
 'payment_preference',
 'aov',
 'repeated']

In [354]:
top_pay =  cust_statistic['payment_preference'].value_counts().head(5).index
sub = cust_statistic[cust_statistic['payment_preference'].isin(top_pay)]
ct = pd.crosstab(sub['payment_preference'],sub['repeated'])

ct

repeated,False,True
payment_preference,,
Easypay,3703,3156
Payaxis,1534,1354
bankalfalah,1177,529
cod,17465,5615
easypay_voucher,1729,2936


In [355]:
chi2, p_val, dof, expected = stt.chi2_contingency(ct)
n = ct.values.sum()
cramers = np.sqrt(chi2/(n*(min(ct.shape)-1)))


print('H0: Payment method and repeat-purchase status are independent')
print('H1: Payment method and repeat-purchase status are associated')
print(ct)
print(f"\nChi2 = {chi2:.1f}, dof = {dof}, p-value = {p_val:.2e}, alpha = 0.05")
print(f"Cramér's V (for effectivity) = {cramers:.3f}")

print(f"İn conclusion: {'Reject H0' if p_val < 0.05 else 'Fail to reject H0'} — statistically significant, "
      f"effect size is {'small-to-moderate' if cramers < 0.3 else 'moderate-to-large'} (V={cramers:.2f}).")


H0: Payment method and repeat-purchase status are independent
H1: Payment method and repeat-purchase status are associated
repeated            False  True 
payment_preference              
Easypay              3703   3156
Payaxis              1534   1354
bankalfalah          1177    529
cod                 17465   5615
easypay_voucher      1729   2936

Chi2 = 3325.3, dof = 4, p-value = 0.00e+00, alpha = 0.05
Cramér's V (for effectivity) = 0.291
İn conclusion: Reject H0 — statistically significant, effect size is small-to-moderate (V=0.29).


A strong statistical association was found between payment method and repeat-purchase behavior `(Chi2=3325.3, p<0.001)`, with Cramér's V=0.29, indicating a moderate association.

Repeat-purchase rates vary significantly by payment method:

- `easypay_voucher: 62.9% (highest)`
- `Payaxis: 46.9%`
- `Easypay: 46.0%`
- `bankalfalah: 31.0%`
- `cod: 24.3% (lowest; largest group — 23,080 customers)`

Conclusion: cod users form the largest segment but have the lowest repeat rate, while `easypay_voucher` users show the `highest loyalty`. This is an association, not causation. Encouraging cod users to adopt alternative payment methods could be a potential retention strategy.

### Region vs AOV

In [356]:
cust_statistic.columns.to_list()

['cust_id',
 'total_revenue',
 'num_orders',
 'total_quantity',
 'num_of_skus',
 'region',
 'gender',
 'payment_preference',
 'aov',
 'repeated']

In [357]:
region_groups = [
    cust_statistic.loc[
        cust_statistic['region'] == region, 'aov'
    ].dropna()
    for region in cust_statistic['region'].dropna().unique()
]

statistic, p_value = stt.kruskal(*region_groups)
print("H0: AOV distribution is the same across all regions")
print("H1: at least one region's AOV distribution differs")
print(f"H-statistic = {statistic:.3f}, p-value = {p_value:.3f}, alpha = 0.05")
print(f"Conclusion: {'Reject H0' if p_value < 0.05 else 'Fail to reject H0'} — "
      f"no statistically significant difference in AOV across regions.")

print(f'H-statistic: {statistic:.3f}')
print(f'p-value: {p_value:.4g}')

H0: AOV distribution is the same across all regions
H1: at least one region's AOV distribution differs
H-statistic = 1.083, p-value = 0.781, alpha = 0.05
Conclusion: Fail to reject H0 — no statistically significant difference in AOV across regions.
H-statistic: 1.083
p-value: 0.7812


### Gender vs Total Revenue 

In [358]:
revenue_male = cust_statistic.loc[
    cust_statistic['gender'] == 'M', 'total_revenue'
].dropna()

revenue_female = cust_statistic.loc[
    cust_statistic['gender'] == 'F', 'total_revenue'
].dropna()

In [359]:
rev_m = cust_statistic.loc[cust_statistic['gender'] == 'M', 'total_revenue'].dropna()
rev_f = cust_statistic.loc[cust_statistic['gender'] == 'F', 'total_revenue'].dropna()

u_stat, p_val3 = stt.mannwhitneyu(rev_m, rev_f, alternative='two-sided')

print("H0: total revenue distribution is the same for male and female customers")
print("H1: total revenue distributions differ by gender")
print(f"n(M) = {len(rev_m):,}, n(F) = {len(rev_f):,}")
print(f"Median revenue — M: {rev_m.median():,.0f}, F: {rev_f.median():,.0f}")
print(f"U = {u_stat:,.0f}, p-value = {p_val3:.3f}, alpha = 0.05")
print(f"Conclusion: {'Reject H0' if p_val3 < 0.05 else 'Fail to reject H0'} — "
      f"no statistically significant difference in spending by gender.")

H0: total revenue distribution is the same for male and female customers
H1: total revenue distributions differ by gender
n(M) = 21,125, n(F) = 21,049
Median revenue — M: 620, F: 620
U = 222,661,510, p-value = 0.791, alpha = 0.05
Conclusion: Fail to reject H0 — no statistically significant difference in spending by gender.


## Correlation Analysis

In [360]:
corr_col = ['total_revenue', 'num_orders', 'total_quantity',
       'num_of_skus','aov']

In [361]:
corr_matrix = cust_statistic[corr_col].corr(method='spearman')

fig_corr = px.imshow(
    corr_matrix.round(2),
    text_auto=True,
    color_continuous_scale='RdBu_r',
    zmin=-1,
    zmax=1,
    title='Spearman Correlation — Customer-Level Metrics'
)

In [362]:

fig_corr.update_layout(height=520)
fig_corr.show()

print(
    f"num_orders vs total_revenue: "
    f"rho = {corr_matrix.loc['num_orders', 'total_revenue']:.2f}"
)

print(
    f"AOV vs total_revenue: "
    f"rho = {corr_matrix.loc['aov', 'total_revenue']:.2f}"
)

print(
    "Correlation does not imply causation — these relationships describe "
    "association within the observed data only."
)

num_orders vs total_revenue: rho = 0.60
AOV vs total_revenue: rho = 0.94
Correlation does not imply causation — these relationships describe association within the observed data only.


## Seasonality

In [363]:
season = orders_cleaned.groupby('order_month').agg(
    revenue = ('revenue','sum'),
    orders = ('order_id','nunique'),
    customers = ('cust_id','nunique')
).reset_index()

In [364]:
season['aov'] = season['revenue'] / season['orders']

season['order_month'] =season['order_month'].astype(str)

In [365]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=season['order_month'],
    y=season['orders'],
    name='Orders',
    marker_color='#2E86AB'
))

fig.add_trace(go.Scatter(
    x=season['order_month'],
    y=season['aov'],
    name='AOV',
    mode='lines+markers',
    line=dict(color='#F26419', width=3),
    yaxis='y2'
))

fig.update_layout(
    title='Monthly Orders and AOV',
    xaxis_title='Month',
    yaxis_title='Orders',
    yaxis2=dict(title='AOV', overlaying='y', side='right'),
    template='plotly_white',
    height=460
)

fig.show()

## Anomaly Detection

In [366]:
season['orders_z'] = (
    season['orders'] - season['orders'].mean()
) / season['orders'].std()

anomalies = season[season['orders_z'].abs() > 0.8]

print("Unusual months:")
display(anomalies[['order_month', 'orders', 'orders_z']].round(2))

fig_anom = px.bar(
    season,
    color=season['orders_z'] > 0,
    color_discrete_map={True: '#2E86AB', False: '#F26419'},
    x='order_month',
    y='orders_z',
    title='Monthly Order Volume Anomalies',
    labels={'orders_z': 'Z-score', 'order_month': 'Month'}
)

fig_anom.add_hline(y=0.8, line_dash='dash')
fig_anom.add_hline(y=-0.8, line_dash='dash')
fig_anom.show()

Unusual months:


,order_month,orders,orders_z
2,2020-12,30153,2.82
6,2021-04,14844,0.89


## Outlier Analysis

In [367]:
q1, q3 = cust_statistic['total_revenue'].quantile([0.25, 0.75])

iqr = q3 - q1
upper = q3 + 1.5 * iqr
outliers = cust_statistic[cust_statistic['total_revenue'] > upper]

print(f"IQR upper side (total_revenue): {upper:,.0f}")
print(f"Outlier customers: {len(outliers):,} ({len(outliers)/len(cust_statistic):.1%} of customers), "
      f"contributing {outliers['total_revenue'].sum()/cust_statistic['total_revenue'].sum():.1%} of total revenue")
print(f"Highest single customer: {cust_statistic['total_revenue'].max():,.0f} revenue across "
      f"{cust_statistic.loc[cust_statistic['total_revenue'].idxmax(), 'num_orders']} orders")

fig_out = px.box(cust_statistic, y='total_revenue', points='outliers',
                  title='Customer Total Revenue — Outlier Check (IQR method)')
fig_out.update_layout(height=450)
fig_out.show()

IQR upper side (total_revenue): 7,307
Outlier customers: 5,951 (14.1% of customers), contributing 74.8% of total revenue
Highest single customer: 557,435 revenue across 23 orders


# Dashboard

In [380]:
# ---------- Slicers ----------
region_options = ['All'] + sorted(orders_cleaned['region'].dropna().unique().tolist())
category_options = ['All'] + sorted(orders_cleaned['category'].dropna().unique().tolist())

region_slicer = widgets.Dropdown(options=region_options, value='All', description='Region:')
category_slicer = widgets.Dropdown(options=category_options, value='All', description='Category:')

# ---------- KPI card template ----------
def kpi_card(title, value, color='#2E86AB'):
    return f"""
    <div style="background:{color};border-radius:10px;padding:18px 26px;text-align:center;
                color:white;min-width:170px;">
        <div style="font-size:13px;opacity:0.85;">{title}</div>
        <div style="font-size:24px;font-weight:bold;margin-top:4px;">{value}</div>
    </div>
    """

kpi_output = widgets.HTML()

# ---------- Chart placeholders (smaller height so all 5 fit on screen) ----------
CHART_HEIGHT = 300

fig_trend = go.FigureWidget()
fig_category = go.FigureWidget()
fig_region = go.FigureWidget()
fig_segment = go.FigureWidget()
fig_payment = go.FigureWidget()

for f in [fig_trend, fig_category, fig_region, fig_segment, fig_payment]:
    f.layout.width = 600
    f.layout.height = CHART_HEIGHT

fig_payment.layout.width = 800
fig_trend.layout.width = 900
# ---------- Update function (runs on every slicer change) ----------
def update_dashboard(*args):
    dff = orders_cleaned.copy()
    if region_slicer.value != 'All':
        dff = dff[dff['region'] == region_slicer.value]
    if category_slicer.value != 'All':
        dff = dff[dff['category'] == category_slicer.value]

    # ---- KPI 1-5 ----
    total_revenue = dff['revenue'].sum()
    total_orders = dff['order_id'].nunique()
    total_customers = dff['cust_id'].nunique()
    aov = total_revenue / total_orders if total_orders else 0
    orders_per_cust = dff.groupby('cust_id')['order_id'].nunique()
    repeat_rate = (orders_per_cust > 1).mean() * 100 if len(orders_per_cust) else 0

    kpi_html = "<div style='display:flex;justify-content:flex-end;gap:22px;flex-wrap:wrap;padding-right:10px;'>"
    kpi_html += kpi_card('Total Revenue', f"{total_revenue:,.0f}", '#2E86AB')
    kpi_html += kpi_card('Total Orders', f"{total_orders:,}", '#F26419')
    kpi_html += kpi_card('Total Customers', f"{total_customers:,}", '#2E86AB')
    kpi_html += kpi_card('AOV', f"{aov:,.1f}", '#F26419')
    kpi_html += kpi_card('Repeat Rate', f"{repeat_rate:.1f}%", '#2E86AB')
    kpi_html += "</div>"
    kpi_output.value = kpi_html

    # ---- Chart 1: Monthly revenue trend ----
    trend = dff.groupby('order_month')['revenue'].sum().reset_index()
    trend['order_month'] = trend['order_month'].astype(str)
    with fig_trend.batch_update():
        fig_trend.data = []
        fig_trend.add_trace(go.Scatter(x=trend['order_month'], y=trend['revenue'],
                                        mode='lines+markers', line=dict(color='#2E86AB', width=3)))
        fig_trend.update_layout(title='Monthly Revenue Trend', xaxis_title='Month',
                                 yaxis_title='Revenue', template='plotly_white',
                                 margin=dict(l=40, r=20, t=40, b=40))

    # ---- Chart 2: Revenue by category ----
    cat = dff.groupby('category')['revenue'].sum().sort_values(ascending=True).reset_index()
    with fig_category.batch_update():
        fig_category.data = []
        fig_category.add_trace(go.Bar(x=cat['revenue'], y=cat['category'],
                                       orientation='h', marker_color='#2E86AB'))
        fig_category.update_layout(title='Revenue by Category', xaxis_title='Revenue',
                                    yaxis_title='', template='plotly_white',
                                    margin=dict(l=110, r=20, t=40, b=40))

    # ---- Chart 3: Revenue by region ----
    reg = dff.groupby('region')['revenue'].sum().sort_values(ascending=False).reset_index()
    with fig_region.batch_update():
        fig_region.data = []
        fig_region.add_trace(go.Bar(x=reg['region'], y=reg['revenue'], marker_color='#F26419'))
        fig_region.update_layout(title='Revenue by Region', xaxis_title='Region',
                                  yaxis_title='Revenue', template='plotly_white',
                                  margin=dict(l=50, r=20, t=40, b=40))

    # ---- Chart 4: Customer segment distribution (RFM) ----
    cust_ids_in_scope = dff['cust_id'].unique()
    seg_scope = rfm[rfm['cust_id'].isin(cust_ids_in_scope)]
    seg_counts = seg_scope['segment'].value_counts().reset_index()
    seg_counts.columns = ['segment', 'customers']
    with fig_segment.batch_update():
        fig_segment.data = []
        fig_segment.add_trace(go.Pie(labels=seg_counts['segment'], values=seg_counts['customers'], hole=0.4))
        fig_segment.update_layout(title='Customer Segment Distribution', template='plotly_white',
                                   margin=dict(l=20, r=20, t=40, b=20))

    # ---- Chart 5: Orders by payment method ----
    pay = dff.groupby('payment_method')['order_id'].nunique().sort_values(ascending=False).reset_index()
    with fig_payment.batch_update():
        fig_payment.data = []
        fig_payment.add_trace(go.Bar(x=pay['payment_method'], y=pay['order_id'], marker_color='#2E86AB'))
        fig_payment.update_layout(title='Orders by Payment Method', xaxis_title='Payment Method',
                                   yaxis_title='Orders', template='plotly_white',
                                   margin=dict(l=50, r=20, t=40, b=60))

# ---------- Wire up slicers ----------
region_slicer.observe(update_dashboard, names='value')
category_slicer.observe(update_dashboard, names='value')

update_dashboard()  # initial render

# ---------- Layout: 3 charts top row, 2 charts bottom row — all 5 visible together ----------
charts_grid = widgets.VBox([
    widgets.HBox([fig_segment, fig_category, fig_region], layout=widgets.Layout(justify_content='flex-start')),
    widgets.HBox([fig_trend, fig_payment], layout=widgets.Layout(justify_content='flex-start'))
])

dashboard = widgets.VBox([
    widgets.HTML("<h2 style='font-family:sans-serif;margin-bottom:4px;'>Sales Performance Dashboard</h2>"),
    widgets.HBox([region_slicer, category_slicer], layout=widgets.Layout(margin='0 0 10px 0')),
    kpi_output,
    charts_grid
])

display(dashboard)

# Final conclusion

## Key Findings
1. **Low completion rate**. Only 49.5% of 286,392 orders were successful, shows that there is a major gap between 'successful' and others.

2. **Extremely low retention**. Weekly retention rates indicates that, customers do not come back after 1st purchase.
> 100% - 6.5% - 6.0% - 3.4% in weeks 0–3. Around 94% of customers do not return after their first purchase.

3. **ARPU increases over time**. The small group of returning customers spends more than other customers, making loyal customers extremely valuable.

4. **Revenue is highly concentrated**. Champions (12.4%) generate 44.5% of revenue, while Champions + Can't Lose Them (21.0%) generate 67.3%.

5. **Category performance is concentrated**. From 15 category only 3 categories gives us 82.3% of revenue. Men's Fashion leads acquisition, while Mobiles & Tablets leads both revenue and repeat purchases (40.3%).

6. **Payment method is linked to loyalty**. Payment method has a high correlation with repeat purchases (p<0.001, Cramér's V=0.29). `cod` has the lowest repeat rate (24.3%), in addition `easypay_voucher` has the highest (62.9%). Region and gender show no  correlation.

7. **Revenue is driven more by AOV than order frequency**. aov has a stronger correlation with total_revenue (ρ=0.94) than number of orders (ρ=0.60).

8. **December 2020 was an exceptional volume spike**, is linked to a seasonal campaign, but the pattern was not repeated.

9. **Geography mainly reflects market size**. `South` region generates 37.6% of revenue and has 37.0% of customers, with similar AOV and retention across regions.

### Business Recommendations

**Acquisition**

- Investigate and replicate the factors behind the December 2020 spike.

**Retention**

- Focus loyalty programs on customers who already demonstrate repeat behavior and higher spending.

**Reactivation**

- Prioritize `At Risk` and `Can't Lose Them` customers for win-back campaigns.

**Customer Value**

- Protect `Champions`, who generate a great share of revenue.

**Product Strategy**

- Use different strategies for acquisition categories and retention categories rather than optimizing solely for revenue.

**Payments**

- Test incentives encouraging `cod` users to adopt higher-retention payment methods, using controlled experiments to establish causality.